File: core/src/apps/common/safety_checks.py
import storage.device as storage_device
from storage.cache_common import APP_COMMON_SAFETY_CHECKS_TEMPORARY
from storage.device import SAFETY_CHECK_LEVEL_PROMPT, SAFETY_CHECK_LEVEL_STRICT
from trezor.enums import SafetyCheckLevel
from trezor.wire import context


def read_setting() -> SafetyCheckLevel:
    """
    Returns the effective safety check level.
    """
    temporary_safety_check_level = context.cache_get(APP_COMMON_SAFETY_CHECKS_TEMPORARY)
    if temporary_safety_check_level:
        return int.from_bytes(temporary_safety_check_level, "big")  # type: ignore [int-into-enum]
    else:
        stored = storage_device.safety_check_level()
        if stored == SAFETY_CHECK_LEVEL_STRICT:
            return SafetyCheckLevel.Strict
        elif stored == SAFETY_CHECK_LEVEL_PROMPT:
            return SafetyCheckLevel.PromptAlways
        else:
            raise ValueError("Unknown SafetyCheckLevel")


def apply_setting(level: SafetyCheckLevel) -> None:
    """
    Changes the safety level settings.
    """
    if level == SafetyCheckLevel.Strict:
        context.cache_delete(APP_COMMON_SAFETY_CHECKS_TEMPORARY)
        storage_device.set_safety_check_level(SAFETY_CHECK_LEVEL_STRICT)
    elif level == SafetyCheckLevel.PromptAlways:
        context.cache_delete(APP_COMMON_SAFETY_CHECKS_TEMPORARY)
        storage_device.set_safety_check_level(SAFETY_CHECK_LEVEL_PROMPT)
    elif level == SafetyCheckLevel.PromptTemporarily:
        storage_device.set_safety_check_level(SAFETY_CHECK_LEVEL_STRICT)
        context.cache_set(APP_COMMON_SAFETY_CHECKS_TEMPORARY, level.to_bytes(1, "big"))
    else:
        raise ValueError("Unknown SafetyCheckLevel")


def is_strict() -> bool:
    """
    Shorthand for checking whether the effective level is Strict.
    """
    return read_setting() == SafetyCheckLevel.Strict

File: sdk/apps/ethereum/src/common/safety_checks.rs
use crate::proto::common::SafetyCheckLevel;

/// Returns the effective safety check level.
pub fn read_setting() -> Result<SafetyCheckLevel> {
    // Check temporary cache first
    // if let Some(cached) = cache_get(APP_COMMON_SAFETY_CHECKS_TEMPORARY) {
    //     if let Some(&byte) = cached.first() {
    //         return SafetyCheckLevel::from_u8(byte);
    //     }
    // }

    // // Fall back to stored setting
    // let stored = storage_device_safety_check_level()?;
    // match stored {
    //     SAFETY_CHECK_LEVEL_STRICT => Ok(SafetyCheckLevel::Strict),
    //     SAFETY_CHECK_LEVEL_PROMPT => Ok(SafetyCheckLevel::PromptAlways),
    //     _ => Err(trezor_app_sdk::Error::InvalidArgument),
    // }
    Ok(SafetyCheckLevel::Strict)
}

/// Changes the safety level settings.
/// TODO not sure if this function should available to ethereum app
pub fn apply_setting(level: SafetyCheckLevel) -> Result<()> {
    // match level {
    //     SafetyCheckLevel::Strict => {
    //         cache_delete(APP_COMMON_SAFETY_CHECKS_TEMPORARY);
    //         storage_device_set_safety_check_level(SAFETY_CHECK_LEVEL_STRICT)?;
    //     }
    //     SafetyCheckLevel::PromptAlways => {
    //         cache_delete(APP_COMMON_SAFETY_CHECKS_TEMPORARY);
    //         storage_device_set_safety_check_level(SAFETY_CHECK_LEVEL_PROMPT)?;
    //     }
    //     SafetyCheckLevel::PromptTemporarily => {
    //         storage_device_set_safety_check_level(SAFETY_CHECK_LEVEL_STRICT)?;
    //         cache_set(APP_COMMON_SAFETY_CHECKS_TEMPORARY, &[level as u8]);
    //     }
    // }
    Ok(())
}

/// Shorthand for checking whether the effective level is Strict.
pub fn is_strict() -> bool {
    read_setting()
        .map(|level| level == SafetyCheckLevel::Strict)
        .unwrap_or(true)
}
